schema/overview.md

```diff
@@ -0,0 +1,68 @@
+IPLD Schemas Overview
+=====================
+
+
+Goals
+-----
+
+1. Provide a reasonable, easy to use, general type system for declaring
+  useful properties about data.
+2. Compose nicely over IPLD Data Model.  Schemas should be usable with any
+  data format you can build an IPLD Codec for.
+3. Be efficient to verify (e.g. predictable; roughly linear in the size of
+  the data and schema; and absolutely not turing complete).
+4. Be language-agnostic: many compatible implementations of the schema
+  checker tools should exist, as well as bindings for every language.
+5. Assist rather than obstruct migration.  (We expect data to exist from
+  *before* the current schemas; we need to work well on this case.)
+6. Be a home for collaboration and documentation!  Schemas should provide a
+  natural place for API documentation, and be a reasonable design literature.
+
+In total: **Make developing decentralized & distributed applications
+easier, clearer, and more robust!**
+
+
+Features
+--------
+
+Well-understood basics (sum types, product types, some specific recursive types):
+
+- typed maps (e.g. `{Foo:Bar}`)
+- typed arrays (e.g. `[Foo]`)
+- typedef'd primitives (e.g. `type Foo int`)
+- structs (e.g. `struct{ f Foo; b Bar }`)
+- discriminated unions (several styles)
+- enumerations (e.g. `["Foo" | "Bar" | "Baz"]`)
+
+Bonus features:
+
+- simple syntax for 'maybe' (e.g. `struct{ f nullable Foo }` or `[nullable Foo]`)
+- clear syntax for non-required fields (e.g. `struct { f optional Foo }`)
+
+All of this maps over the IPLD Data Model:
+
+- every type has a representation;
+- e.g. maps are represented as maps; structs are also represented as maps!
+- representations can specify aliases for how to serialize things; e.g. provide
+  shorter serial strings to use in place of longer struct names.
```

Comment:
ahh, "to use in place", a question I've been meaning to raise about aliases is: does a serialiser use the primary or the alias or is it left up to the implementation because it doesn't matter? is the purpose of an alias to allow the _deserialising_ of data with an alternative form or to allow the _serialising_ of data in an alternate (maybe more efficient?) form, or some undefined combination of both? herein lies both the consistent-hash problem and the consistent-across-implementations problem.

doesn't need to be addressed here, just noting for future discussion

Reply:
**must** use alias.  consistency problem is right -- both for hashing, and also for cardinality correctness even aside from hashing.

Reply:
... so, 'alias' might be an ideal name for this feature :thinking:

Reply:
well the name "alias" made me assume it was an alternate form valid during deserialisation and that serialisation would always use the canonical name that's in the parent `type` section. So if it's a "**must** use alias", then yes, I think it's the wrong term. I don't have a suggestion for an alternative just at the moment though.